### PR TITLE
Terms: Moving category-selector to term-selector

### PR DESCRIPTION
**Pull request by timmyc**
For #3697- This branch begins the task of transitioning the `<CategorySelector />` to a generic `<TermSelector />` for use with CPT taxonomies.  This has branch has been a bit of a winding trail through the redux forest, but it has now arrived at a happy meadow, and is ready for a review or two 👍 

The final path of this branch involved using much of the same logic that @aduth used in #5647.  The `<CategorySelector />` and `<PostSelector />` have shared many similarities in the past, and this continues here with the adoption of ( much of the same code ) and `ReactVirtualized`.  I wonder if there might be any way to possibly merge these two components since they are quite similar.  Though there are differences, and maybe some magic is best left untouched.

Akin to the `<CategorySelector />`, the first implementation of the `<TermSelector />` is in the menu builder - replacing the current usage of `<CategorySelector />` there.  This is the first bit of redux in this area, so that is kind of fun too!

__To Test__
Essentially to test this branch out, it would be most helpful to have a test site with over 100 categories to play with the InfiniteScroll provided by ReactVirtualized.  Additionally, if you have some nested categories, that is desirable as well.

With copious amounts of test categories available, simply work with the Menu Builder Category section in all crazy ways possible:

menus_ _testingtimmy2_wordpress_com_ _wordpress_com

Specifically test out the search capability, and infinite scroll.  But basic actions such as selecting a different category, saving, and ensuring the menu updates properly would be excellent testing steps.

__Questions__
There was some chatter around possibly naming this `<HierarchicalTermSelector />` and having a `<TermSelector />` that was smart enough to determine if this component, or the token field should be rendered.  Naming stuff is hard - so your feedback there is appreciated too.  /cc @nylen

### CODE REVIEW

File: client/my-sites/term-selector/no-results.jsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import PureComponent from 'react-pure-render/component';
+
+/**
+ * Internal dependencies
+ */
+import localize from 'lib/mixins/i18n/localize';
```
Comment (Tug) [inline client/my-sites/term-selector/no-results.jsx]: Hi there, i18n usage has changed recently. please use:
```
import { localize } from 'i18n-calypso';
```
instead.

**Comment by aduth** [comment]
One problem I see is that we're still using the idea of "fetch next page of items when we need more" type logic, which won't work with persisted state*. I've been meaning to work on `<PostSelector />` to change this. It'll need to look more like in #5995, where we determine, based on the range of rendered items, which pages should be requested (looking ahead and behind some). In fact, we'll probably be able to do away with `<InfiniteLoader />` altogether both here and in `<PostSelector />`.

__*__ See "Implementation notes" for #5995

File: client/my-sites/menus/controller.js [hunk 2]
```diff
@@ -3,7 +3,8 @@
  */
 var ReactDom = require( 'react-dom' ),
 	React = require( 'react'),
-	i18n = require( 'i18n-calypso' );
+	i18n = require( 'i18n-calypso' ),
+	ReactRedux = require( 'react-redux' );
```
Comment (aduth) [inline client/my-sites/menus/controller.js]: I think we've been generally okay with mixing import styles, avoiding adding new `require` at all. It's ugly, but I guess you can argue that the ugliness will hasten the transition 😉

**Comment by aduth** [comment]
I'll leave it to your judgment, but I wonder if it'd be safer, allow faster iteration, and move us along our CPT milestone to implement this first in the feature-flagged [editor taxonomies accordions](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-drawer/taxonomies.jsx) instead of trying to tackle menus first?

File: client/my-sites/menus/style.scss [hunk 3]
```diff
@@ -738,6 +738,44 @@
...
+				&.menu-item-options__term-selector {
+					padding: 10px 0;
+					overflow-y: hidden;	
+
+					.term-selector {
+						border: 0;
+					}
+
+					.term-selector__results {
+						&::-webkit-scrollbar {
```
Comment (aduth) [inline client/my-sites/menus/style.scss]: Why are these styles needed? What about other browsers?

File: client/my-sites/term-selector/no-results.jsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import PureComponent from 'react-pure-render/component';
+import { localize } from 'i18n-calypso';
+
+class TermSelectorNoResults extends PureComponent {
```
Comment (aduth) [inline client/my-sites/term-selector/no-results.jsx]: You've mixed `React.createClass` and `Component` classes in this pull request. Any particular reason for this?

Related: p1466013240000130-slack-calypso-framework

File: client/my-sites/term-selector/no-results.jsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	constructor( props ) {
```
Comment (aduth) [inline client/my-sites/term-selector/no-results.jsx]: You don't need to define `constructor` if you're not doing anything within it.

File: client/my-sites/term-selector/index.jsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	incrementPage() {
+		this.setState( {
+			page: this.state.page + 1
+		} );
+	},
+
+	componentWillReceiveProps( nextProps ) {
+		const isChangingQuery = [
+			'taxonomy',
+			'multiple'
```
Comment (aduth) [inline client/my-sites/term-selector/index.jsx]: I don't think `multiple` would necessitate a page reset?

Comment (aduth) [inline client/my-sites/term-selector/index.jsx]: In which case, this entire function could be simplified to a single `nextProps.taxonomy !== this.props.taxonomy` condition.

File: client/my-sites/term-selector/no-results.jsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+		super( props );
+	}
+
+	render() {
+		const { createLink, translate } = this.props;
+		let createMessage;
+		let noResultsMessage = translate( 'No results. Please try a different search.' );
```
Comment (aduth) [inline client/my-sites/term-selector/no-results.jsx]: This should be `const` since you're not reassigning ([guideline](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#variable-declarations)). I'd probably go one step further and simply inline this in the returned element.

Comment (aduth) [inline client/my-sites/term-selector/no-results.jsx]: Now that I mention it, we should really turn on the [ESLint `prefer-const` rule](http://eslint.org/docs/rules/prefer-const).

File: client/my-sites/menus/style.scss (hunk 3 above)
Comment (timmyc) [inline client/my-sites/menus/style.scss]: I'm mimicking the scrollbar logic seen below to try to make the scroll bar look like it does in other menu areas.  I'm fine with removing it.

File: client/my-sites/term-selector/terms.jsx [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+		const onChange = ( ...args ) => this.props.onChange( item, ...args );
+		const setItemRef = ( ...args ) => this.setItemRef( item, ...args );
+
+		const { multiple, defaultTermId, taxonomy, translate } = this.props;
+		const selectedIds = this.getSelectedIds();
+		const itemId = item.ID;
+		const name = unescapeString( item.name ) || translate( 'Untitled' );
+		const checked = includes( selectedIds, itemId );
+		const inputType = multiple ? 'checkbox' : 'radio';
+		const domId = taxonomy + '-option-' + itemId;
```
Comment (aduth) [inline client/my-sites/term-selector/terms.jsx]: Some of this stuff I couldn't find being used in `<PostSelector />` so I removed. Maybe we can use this as an opportunity to revisit whether some of these attributes are necessary (`id`, `name`)?

File: client/state/terms/selectors.js [hunk 9]
```diff
@@ -103,7 +103,16 @@ export const getTermsHierarchyForQueryIgnoringPage = createSelector(
 			return terms;
 		}
 
-		return ( new TreeConvert( 'ID' ) ).treeify( terms );
+		// For each site term object, add `parent` and `order` properties.
+		// These are used by the TreeConvert library ( which mutates ) to build the hierarchy.
+		const treeReadyTerms = terms.map( ( term, i ) => {
+			return Object.assign( {}, term, {
```
Comment (aduth) [inline client/state/terms/selectors.js]: Fancier option now with Babel 6 is [object spread properties](https://github.com/sebmarkbage/ecmascript-rest-spread) (stage 2 proposal):

```js
return {
	...term,
	parent: term.parent || 0,
	order: i
};
```

File: client/my-sites/term-selector/no-results.jsx (hunk 7 above)
Comment (aduth) [inline client/my-sites/term-selector/no-results.jsx]: >Now that I mention it, we should really turn on the [ESLint `prefer-const` rule](http://eslint.org/docs/rules/prefer-const).

#6093

**Comment by nylen** [comment]
I do think this should be called `<HierarchicalTermSelector />` because non-hierarchical terms will need a separate component (perhaps `NonHierarchicalTermSelector`).

**Comment by aduth** [comment]
Since there's not anything technically stopping a hierarchical taxonomy from using a token field and vice-versa, aside from UX considerations, perhaps we should not include this in the name and instead describe the characteristics of the UI?

I'm thinking `<TermTokenFIeld />` and `<TermTreeSelector />`, perhaps eventually a `<TermSelector />` to _magically_ choose to best option to render based on the hierarchical nature of the taxonomy type.

**Comment by timmyc** [comment]
I like `<TermTreeSelector />` here as well.  🌳

**Comment by timmyc** [comment]
General feedback applied in b648e2f, and the renaming took place in 0a30c02.  In 3dc6dc9 the notion of tracking the pages to load akin to #5995.  There @aduth I see you are tracking the first and last page loading - here I created a new selector for `isRequestingTermsForQueryIgnoringPage` to capture if any related request for term data is in-progress.  Let me know if you think that is a good approach.

File: client/my-sites/term-tree-selector/index.jsx [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+	render() {
+		const { className, taxonomy, onChange, selected, createLink } = this.props;
+
+		const classes = classNames( className );
+		const { search } = this.state;
+		const query = { search };
+
+		return (
+			<div className={ classes } ref="wrapper">
```
Comment (aduth) [inline client/my-sites/term-tree-selector/index.jsx]: What's the `ref` used for?

File: client/my-sites/term-tree-selector/index.jsx [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	getInitialState() {
+		return {
+			search: ''
+		};
+	},
+
+	onSearch( searchTerm ) {
+		if ( searchTerm !== this.state.search ) {
+			this.setState( {
+				page: 1,
```
Comment (aduth) [inline client/my-sites/term-tree-selector/index.jsx]: `page` no longer exists in `state`

File: client/my-sites/term-tree-selector/index.jsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import TermTreeSelectorTerms from './terms';
```
Comment (aduth) [inline client/my-sites/term-tree-selector/index.jsx]: I might suggest looking to see if we can eliminate `./terms` as a separate component altogether and merge into this one. I think it was a problem in `<PostSelector />` only because of the need to compile a query object and manage page in state.

File: client/my-sites/term-tree-selector/no-results.jsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import PureRenderMixin from 'react-pure-render/mixin';
+
+export default React.createClass( {
```
Comment (aduth) [inline client/my-sites/term-tree-selector/no-results.jsx]: What was the reason for moving away from `Component` class to `React.createClass`? Personally I'm strongly in favor of discouraging `React.createClass`.

File: client/my-sites/term-tree-selector/no-results.jsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+		createLink: PropTypes.string
+	},
+
+	render() {
+		const { createLink } = this.props;
+		let createMessage;
+
+		if ( createLink ) {
+			createMessage = this.translate( 'You may want to {{a}}create a new category{{/a}}.', {
+				context: 'Menus: item search/listing results',
```
Comment (aduth) [inline client/my-sites/term-tree-selector/no-results.jsx]: Not specific to menus

File: client/my-sites/term-tree-selector/terms.jsx [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+						siteId={ siteId }
+						taxonomy={ taxonomy }
+						query={ { ...query, page } } />
+				) ) }
+				{ showSearch && (
+					<Search
+						searchTerm={ this.state.searchTerm }
+						onSearch={ this.onSearch } />
+				) }
+				<InfiniteLoader
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: Why do you still need `<InfiniteLoader />` with `VirtualScroll`'s `onRowsRendered` handling setting pages to be requested?

File: client/my-sites/term-tree-selector/terms.jsx [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+					) }
+				</InfiniteLoader>
+			</div>
+		);
+	}
+} );
+
+export default connect( ( state, ownProps ) => {
+	const siteId = getSelectedSiteId( state );
+	const { taxonomy, search, query } = ownProps;
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: `search` is not used by any of the selectors below and can be omitted from the `connect` selector.

File: client/my-sites/term-tree-selector/terms.jsx [hunk 17]
```diff
@@ -0,0 +1,422 @@
...
+
+		return (
+			<div key="placeholder" className="term-tree-selector__list-item is-placeholder">
+				<label>
+					<input
+						type={ this.props.multiple ? 'checkbox' : 'radio' }
+						disabled
+						className="term-tree-selector__input" />
+					<span className="term-tree-selector__label">
+						{ this.translate( 'Loading…' ) }
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: We should be consistent with our usage of `localize` vs. the `this.translate` mixin, probably changing this to `this.props.translate`

File: client/my-sites/term-tree-selector/terms.jsx [hunk 18]
```diff
@@ -0,0 +1,422 @@
...
+
+		const { multiple, defaultTermId, translate } = this.props;
+		const selectedIds = this.getSelectedIds();
+		const itemId = item.ID;
+		const name = unescapeString( item.name ) || translate( 'Untitled' );
+		const checked = includes( selectedIds, itemId );
+		const inputType = multiple ? 'checkbox' : 'radio';
+		let disabled;
+
+		if ( multiple && checked && defaultTermId &&
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: Could probably move this into the default assigned value for `disabled`?

```js
const disabled = (
	multiple &&
	checked &&
	defaultTermId &&
	1 === selectedIds.length &&
	defaultTermId === itemId	
);
```

File: client/my-sites/term-tree-selector/no-results.jsx (hunk 13 above)
Comment (timmyc) [inline client/my-sites/term-tree-selector/no-results.jsx]: Just based off your consistency comment.  Old habits.

File: client/my-sites/term-tree-selector/index.jsx (hunk 10 above)
Comment (timmyc) [inline client/my-sites/term-tree-selector/index.jsx]: remnant from the old category-selector :)

File: client/my-sites/term-tree-selector/index.jsx (hunk 12 above)
Comment (timmyc) [inline client/my-sites/term-tree-selector/index.jsx]: I think we still need this setup to manage `search` coming in to `./terms` as part of the `query` prop.

**Comment by timmyc** [comment]
@aduth incorporated your feedback in 6d8eda8.  I believe we still need the parent component to create the `query` to pass into `terms.jsx` to properly use the selectors.  I attempted to consolidate into a single component but that was a limiting factor for me.

I did however simplify the `rowRender` by moving placeholder and no-results logic into a `noRowsRenderer` method.

**Comment by aduth** [comment]
Looks like we've lost the back button on mobile:

Back missing

(Other tabs show e.g. "< Posts" above options)

File: client/my-sites/term-tree-selector/terms.jsx [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+				{ item.items && (
+					<div className="term-tree-selector__nested-list">
+						{ item.items.map( this.renderItem ) }
+					</div>
+				) }
+			</div>
+		);
+	},
+
+	renderNoResults() {
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: Using `noRowsRendered` seems pretty clever, and really helps simplify the problem of managing the placeholder as being considered in the row count. Only downside I can think of is that we lose the option to show a placeholder when "refreshing" a search term, e.g. search "Foo", clear, then search "Foo" again. Previous search results will be shown and we load latest results in the background, but no indication we're doing so. Not sure this is a blocker.

**Comment by aduth** [comment]
There's a bit of styling difference between master and this branch, not sure if desirable:

wpcalypso|try/category-selector-to-terms
---|---
Before|After

**Comment by aduth** [comment]
Found a bug, though I think it may apply to menus generally, not specific to these changes.

1. Add a menu item
2. Remove added menu item by clicking Edit, then Delete
3. Click Cancel in menu save bar
4. Note that you can no longer change menus

File: client/my-sites/term-tree-selector/terms.jsx [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+	},
+
+	renderItem( item ) {
+		const onChange = ( ...args ) => this.props.onChange( item, ...args );
+		const setItemRef = ( ...args ) => this.setItemRef( item, ...args );
+
+		const { multiple, defaultTermId, translate } = this.props;
+		const selectedIds = this.getSelectedIds();
+		const itemId = item.ID;
+		const name = unescapeString( item.name ) || translate( 'Untitled' );
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: `decodeEntities`?

File: client/my-sites/term-tree-selector/terms.jsx [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+		if ( ! this.hasPerformedSearch ) {
+			this.hasPerformedSearch = true;
+			analytics.ga.recordEvent( this.props.analyticsPrefix, 'Performed Term Search' );
+		}
+
+		this.setState( { searchTerm } );
+		this.debouncedSearch();
+	},
+
+	getSelectedIds( selected ) {
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: Since argument is not used, should probably be left out to avoid confusion.

File: client/my-sites/term-tree-selector/terms.jsx [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+
+		this.queueRecomputeRowHeights = debounce( this.recomputeRowHeights );
+		this.debouncedSearch = debounce( () => {
+			this.props.onSearch( this.state.searchTerm );
+		}, SEARCH_DEBOUNCE_TIME_MS );
+	},
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.taxonomy !== this.props.taxonomy ) {
+			this.setState( {
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: Could get away with `this.setState( this.getInitialState() )` here but sometimes can lead to maintenance issues if someone revises `getInitialState` without understanding the repercussions.

File: client/my-sites/term-tree-selector/terms.jsx [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+				searchTerm: '',
+				requestedPages: [ 1 ]
+			} );
+		}
+	},
+
+	componentDidUpdate( prevProps ) {
+		const forceUpdate = (
+			prevProps.selected !== this.props.selected ||
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: If `selected` is array, can we do strict equality here? `[ 100 ] !== [ 100 ]`

**Comment by aduth** [comment]
Getting close, for sure. Looking good. Mobile regression definitely needs a fix, though.

**Comment by timmyc** [comment]
I'm having a hard time recreating [this issue](https://github.com/Automattic/wp-calypso/pull/5366#issuecomment-227142849) - does it happen on production for you as well @aduth ?

**Comment by aduth** [comment]
@timmyc : Yeah, I seem to be having the same issues in production, so definitely not specific to this branch. Here's a screencast: https://cloudup.com/coVP-2paoEO

**Comment by timmyc** [comment]
In af003d7 I got the styling much closer to what currently is in place in the menu category panel.  Still some discrepancies primarily due to the scrolling container changing.  Though I did update the `showSearch` logic to match how things flow in the `<CategorySelector />` now - where the search box is not shown until the terms are loaded and it is needed/not compact.

File: client/my-sites/term-tree-selector/terms.jsx [hunk 24]
```diff
@@ -0,0 +1,406 @@
...
+				</label>
+			</div>
+		);
+	},
+
+	render() {
+		const rowCount = this.getRowCount();
+		const isCompact = this.isCompact();
+		const searchLength = this.state.searchTerm.length;
+		const showSearch = ( searchLength > 0 || ! isCompact ) &&
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: Can you explain why `searchLength > 0 || ! isCompact` wasn't sufficient here?

**Comment by aduth** [comment]
I'm observing that I must click an option twice for it to become selected. The first time, it's focussed, but not selected.

select2

**Comment by aduth** [comment]
Fine to address separately, but we should also add this to [`devdocs/app-components`](https://wpcalypso.wordpress.com/devdocs/app-components) (currently experiencing some difficulties, #6205)

File: client/state/terms/test/selectors.js [hunk 25]
```diff
@@ -65,6 +66,66 @@ describe( 'selectors', () => {
 		} );
 	} );
 
+	describe( 'isRequestingTermsForQueryIgnoringPage()', () => {
+		it( 'should return false if no request exists', () => {
+			const requesting = isRequestingTermsForQueryIgnoringPage( {
+				terms: {
+					queryRequests: {}
+				}
+			}, 2916284, 'categories', {} );
```
Comment (aduth) [inline client/state/terms/test/selectors.js]: Noticed in #6224 that most tests are specifying `categories` as the taxonomy, when the proper name should be `category` (technically, there could be another taxonomy named `categories`, but I assume this refers to the one for posts). I updated the affected tests in #6224 (rebased commit).

https://public-api.wordpress.com/rest/v1.1/sites/en.blog.wordpress.com/post-types/post/taxonomies

File: client/my-sites/term-tree-selector/terms.jsx (hunk 24 above)
Comment (timmyc) [inline client/my-sites/term-tree-selector/terms.jsx]: The additional condition is to prevent the search box from being shown on the initial load - just like it currently acts.

File: client/state/terms/test/selectors.js (hunk 25 above)
Comment (timmyc) [inline client/state/terms/test/selectors.js]: Yes I have noticed this, corrected it in some places, and left it in others ( like here ) - since as you said potentially there could be a custom `categories` taxonomy.

**Comment by timmyc** [comment]
Fixed the selection issue in c7a7160 ( derp ).  Still thinking of bringing this in after #6224 lands @aduth - so hopefully we can get that out 👍

**Comment by timmyc** [comment]
@aduth found a few issues with the QueryTermManager upon rebase.  Specifically the `TERMS_RECEIVE` action object was missing the `found` attribute which was messing up pagination.  Also, I had to specify a `number: 100` as a default in the query.  You can see I am doing this in `terms.jsx` - wasn't sure if I should be doing this in a different manner elsewhere... perhaps in `actions` instead?

With those changes things seem to be working well now.  The one oddity I am seeing is when after searching, and then hitting `esc` or the (x) to clear the search term, the term selector list is momentarily rendered out of order/un-treeified.

**Comment by timmyc** [comment]
> The one oddity I am seeing is when after searching, and then hitting esc or the (x) to clear the search term, the term selector list is momentarily rendered out of order/un-treeified.

I'm no longer seeing this issue now. 

@aduth if you could review 9b8668b and 0bf998d I would appreciate it 👍

File: client/my-sites/term-tree-selector/terms.jsx [hunk 26]
```diff
@@ -0,0 +1,407 @@
...
+					className="term-tree-selector__results" />
+			</div>
+		);
+	}
+} );
+
+export default connect( ( state, ownProps ) => {
+	const siteId = getSelectedSiteId( state );
+	const { taxonomy, query } = ownProps;
+	query.number = query.number || 100;
```
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: Why did you need to set a fallback of `100` here?

File: client/state/terms/selectors.js [hunk 27]
```diff
@@ -26,6 +27,28 @@ export function isRequestingTermsForQuery( state, siteId, taxonomy, query ) {
...
+ * Returns true if currently requesting terms for a query, excluding all known
+ * queried pages, or false otherwise.
+ *
+ * @param  {Object}  state    Global state tree
+ * @param  {Number}  siteId   Site ID
+ * @param  {String}  taxonomy Taxonomy slug
+ * @param  {Object}  query    Terms query object
+ * @return {Boolean}           Terms for the query
+ */
+export function isRequestingTermsForQueryIgnoringPage( state, siteId, taxonomy, query ) {
```
Comment (aduth) [inline client/state/terms/selectors.js]: You should use `TermQueryManager#getItemsIgnoringPage` to implement this ([see equivalent posts implementation](https://github.com/Automattic/wp-calypso/blob/257ec0c69d25d750c02af22eec409676f685daf5/client/state/posts/selectors.js#L171-L194)). You can drop `range` dependency after too.

File: client/state/terms/actions.js [hunk 28]
```diff
@@ -35,7 +35,7 @@ export function addTerm( siteId, taxonomy, term ) {
 		} ) );
 
 		return wpcom.site( siteId ).taxonomy( taxonomy ).term().add( term ).then(
-			( data ) => dispatch( receiveTerm( siteId, taxonomy, omit( data, '_headers' ) ) ),
+			( data ) => dispatch( receiveTerm( siteId, taxonomy, omit( data, '_headers' ) ), null, null ),
```
Comment (aduth) [inline client/state/terms/actions.js]: What's `null, null` for? `dispatch` only [takes a single argument](http://redux.js.org/docs/api/Store.html#dispatch)

File: client/state/terms/actions.js [hunk 29]
```diff
@@ -109,13 +111,14 @@ export function requestSiteTerms( siteId, taxonomy, query = {} ) {
 		} );
 
 		return wpcom.site( siteId ).taxonomy( taxonomy ).termsList( query ).then( ( data ) => {
-			dispatch( receiveTerms( siteId, taxonomy, data.terms, query ) );
 			dispatch( {
 				type: TERMS_REQUEST_SUCCESS,
+				found: data.found,
```
Comment (aduth) [inline client/state/terms/actions.js]: Why does `TERMS_REQUEST_SUCCESS` need `found`?

File: client/my-sites/term-tree-selector/terms.jsx (hunk 26 above)
Comment (timmyc) [inline client/my-sites/term-tree-selector/terms.jsx]: `QueryManager` has a default of `20`.  Without setting `100`, the term endpoint default, pagination and all else goes sideways.

File: client/state/terms/actions.js (hunk 28 above)
Comment (timmyc) [inline client/state/terms/actions.js]: doh, supposed to be on the call to `receiveTerm`

File: client/my-sites/term-tree-selector/terms.jsx (hunk 26 above)
Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: >`QueryManager` has a default of `20`.  Without setting `100`, the term endpoint default, pagination and all else goes sideways.

Hmm, that shouldn't be the case for `TermQueryManager`. We override the default query and specify the correct value of `100`. If you're seeing 20, sounds like a problem with the query manager implementation.

https://github.com/Automattic/wp-calypso/blob/257ec0c/client/lib/query-manager/term/index.js#L70
https://github.com/Automattic/wp-calypso/blob/257ec0c/client/lib/query-manager/term/constants.js#L5

File: client/state/terms/actions.js (hunk 28 above)
Comment (aduth) [inline client/state/terms/actions.js]: >doh, supposed to be on the call to `receiveTerm`

Do you need those though?

File: client/state/terms/selectors.js (hunk 27 above)
Comment (timmyc) [inline client/state/terms/selectors.js]: Actually this is to determine if there are any active taxonomy/term requests regardless of page for a given query - so we are looking at `state.terms.queryRequests` not the items.

File: client/state/terms/actions.js (hunk 29 above)
Comment (timmyc) [inline client/state/terms/actions.js]: It does not, removing.

File: client/state/terms/selectors.js (hunk 27 above)
Comment (aduth) [inline client/state/terms/selectors.js]: Ah, whoops, for some reason thought this was `getTermsForQueryIgnoringPage`

Comment (timmyc) [inline client/state/terms/selectors.js]: `tooManyNearlyIdenticalLongSelectorNames`

File: client/my-sites/term-tree-selector/terms.jsx (hunk 26 above)
Comment (timmyc) [inline client/my-sites/term-tree-selector/terms.jsx]: It appears the issue is due to these lines: https://github.com/Automattic/wp-calypso/blob/257ec0c69d25d750c02af22eec409676f685daf5/client/lib/query-manager/paginated/index.js#L150-L151

changing to:

```
const page = options.query.page || this.constructor.DEFAULT_QUERY.page;
const perPage = options.query.number || this.constructor.DEFAULT_QUERY.number;
```
solves the issue.

Should I fix that in this branch here, or leave my work-around in place and circle back in a subsequent PR?

Comment (aduth) [inline client/my-sites/term-tree-selector/terms.jsx]: >changing to:

Nice catch! Looks like the right solution. I'd be fine with including it in these changes. Wonder if it'd not be too much more work to try to write a (currently-failing) test to reflect the need for that specific change.

File: client/my-sites/term-tree-selector/README.md [hunk 30]
```diff
@@ -0,0 +1,104 @@
+TermTreeSelector
```
Comment (aduth) [inline client/my-sites/term-tree-selector/README.md]: @mtias Any thoughts on having this component in `my-sites`, given discussions around app components? (In any case, I think it'd be fine to move after merging this one)

**Comment by aduth** [comment]
Did a final pass and round of testing. Looking good and behaving well (loving the effect of persistence on reloads and multi-query tracking on clearing searches). 

Could use a bit of squashing but otherwise looking good to merge 👍  Glad to see this getting in.
